R/util.R

```diff
@@ -145,6 +145,10 @@ expr_sum <- function(x) {
 
 
 expr_prod <- function(x) {
+  if (rlang::is_call(x, "dim")) {
```

Comment:
This should not be needed, what fails when it goes away

Reply:
I get failures like this:-
```
Error (test-generate.R:946:3): can generate system with aliased array
Error in `for (el in x) {
    ret <- times(ret, el)
}`: invalid for() loop sequence
Backtrace:
     ▆
  1. └─odin2:::odin_parse(...) at test-generate.R:946:3
  2.   └─odin2:::odin_parse_quo(...) at odin2/R/parse.R:3:3
  3.     └─odin2:::parse_system_overall(exprs, call) at odin2/R/parse.R:15:3
  4.       ├─odin2:::resolve_array_references(...) at odin2/R/parse_system.R:135:3
  5.       │ └─odin2:::vlapply(arrays$dims, rlang::is_call, "dim") at odin2/R/parse_system.R:246:3
  6.       │   └─base::vapply(..., FUN.VALUE = TRUE) at odin2/R/util.R:12:3
  7.       ├─odin2:::data_frame(...) at odin2/R/parse_system.R:135:3
  8.       │ └─base::data.frame(..., stringsAsFactors = FALSE, check.names = FALSE) at odin2/R/util.R:79:3
  9.       ├─base::I(lapply(dims, expr_prod)) at odin2/R/parse_system.R:135:3
 10.       │ └─base::unique.default(c("AsIs", oldClass(x)))
 11.       └─base::lapply(dims, expr_prod) at odin2/R/parse_system.R:135:3
 12.         └─odin2 (local) FUN(X[[i]], ...)
 ```
 
 which ends up being in parse_system.R:136-141 - 
 
 ```
   arrays <- resolve_array_references(data_frame(
    name = vcapply(exprs[is_dim], function(x) x$lhs$name_data),
    alias = rep(NA_character_, length(dims)),
    rank = lengths(dims),
    dims = I(dims),
    size = I(lapply(dims, expr_prod))))
```

the size definition, where dims is something like:
```
[[1]]
[[1]][[1]]
[1] 1


[[2]]
dim(a)
```

Reply:
Effectively: 
```
expr_prod(quote(dim(a)))
```